lib/vtls/schannel.c

```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
+     */
+    cert_hostname_buff = (LPTSTR)malloc(len * sizeof(TCHAR));
+    actual_len = CertGetNameString(pCertContextServer,
+                                   CERT_NAME_DNS_TYPE,
+                                   name_flags,
+                                   NULL,
+                                   (LPTSTR) cert_hostname_buff,
+                                   len);
+
+    /* Sanity check */
+    if(actual_len != len) {
+      failf(data,
+            "schannel: CertGetNameString() returned certificate "
+            "name information of unexpected size");
+      result = CURLE_PEER_FAILED_VERIFICATION;
+      break;
+    }
+
+    /* If HAVE_CERT_NAME_SEARCH_ALL_NAMES is available, the output
+     * will contain all DNS names, where each name is null-terminated
+     * and the last DNS name is double null-terminated. Due to this
+     * encoding, use the length of the buffer to iterate over all names.
+     */
+    result = CURLE_PEER_FAILED_VERIFICATION;
+    while(cert_hostname_buff_index < len
+          && cert_hostname_buff[cert_hostname_buff_index] != TEXT('\0')
+          && result == CURLE_PEER_FAILED_VERIFICATION) {
+
+      const char *cert_hostname;
+
+      /* Comparing the cert name and the connection hostname encoded as UTF-8
+       * is acceptable since both values are assumed to use ASCII
+       * (or some equivalent) encoding
+       */
+      cert_hostname = Curl_convert_tchar_to_UTF8(
+          &cert_hostname_buff[cert_hostname_buff_index]);
+      if(!cert_hostname) {
+        result = CURLE_OUT_OF_MEMORY;
+      }
+      else{
+        int match_result;
+
+        match_result = Curl_cert_hostcheck(cert_hostname, conn->host.name);
+        if(match_result == CURL_HOST_MATCH) {
+          infof(data,
+              "schannel: connection hostname (%s) validated "
+              "against certificate name (%s)\n",
+              conn->host.name,
+              cert_hostname);
+          result = CURLE_OK;
+        }
+        else{
+          size_t cert_hostname_len;
+
+          infof(data,
+                "schannel: connection hostname (%s) did not match "
+                "against certificate name (%s)\n",
+                conn->host.name,
+                cert_hostname);
+
+          cert_hostname_len = _tcslen(
+              &cert_hostname_buff[cert_hostname_buff_index]);
+
+          /* Move on to next cert name */
+          cert_hostname_buff_index += cert_hostname_len + 1;
+
+          result = CURLE_PEER_FAILED_VERIFICATION;
+        }
+        Curl_unicodefree(cert_hostname);
+      }
+    }
+
+    if(result == CURLE_PEER_FAILED_VERIFICATION) {
+      failf(data,
+            "schannel: CertGetNameString() failed to match "
+            "connection hostname (%s) against server certificate names",
+            conn->host.name);
+    }
+    else if(result != CURLE_OK) {
+      failf(data, "schannel: server certificate name verification failed");
+    }
+  }while(0);
```

Comment:
Same as before. I think you are doing this to break out and go to the cleanup part of the function. But in this case I would actually recommend to use goto, since this is used in libcurl for such purposes and actually describes what you are doing.